race detected during execution of test

<!-- Make sure that you visit our User Guide at https://kubevirt.io/user-guide.
-->

**What happened**:
when run command `KUBEVIRT_RUN_UNNESTED=true make go-test` failed with data race.
```sh
=== RUN   TestRest
Running Suite: pkg/virt-api/rest/rest_suite_test.go - /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest
====================================================================================================================
Random Seed: 1668130577

Will run 217 of 217 specs
•••••••••••••••==================
WARNING: DATA RACE
Write at 0x00c000396a18 by goroutine 47:
  kubevirt.io/kubevirt/pkg/virt-api/rest.glob..func3.1()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:50 +0x1e4
  github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func2()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:605 +0xe7

Previous read at 0x00c000396a18 by goroutine 62:
  kubevirt.io/kubevirt/pkg/virt-api/rest.glob..func3.1.4()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:68 +0x4e
  kubevirt.io/kubevirt/pkg/virt-api/rest.(*Streamer).Handle.func4()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer.go:99 +0x74

Goroutine 47 (running) created at:
  github.com/onsi/ginkgo/v2/internal.(*Suite).runNode()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:593 +0x8f9
  github.com/onsi/ginkgo/v2/internal.(*group).attemptSpec()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/group.go:181 +0x2c98
  github.com/onsi/ginkgo/v2/internal.(*group).run()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/group.go:303 +0xb9b
  github.com/onsi/ginkgo/v2/internal.(*Suite).runSpecs()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:304 +0xe71
  github.com/onsi/ginkgo/v2/internal.(*Suite).Run()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:83 +0x424
  github.com/onsi/ginkgo/v2.RunSpecs()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/core_dsl.go:280 +0x1044
  kubevirt.io/client-go/testutils.KubeVirtTestSuiteSetup()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/kubevirt.io/client-go/testutils/setup.go:55 +0x6a5
  kubevirt.io/kubevirt/pkg/virt-api/rest.TestRest()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/rest_suite_test.go:29 +0x2e
  testing.tRunner()
      /root/sdk/go1.18.5/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /root/sdk/go1.18.5/src/testing/testing.go:1486 +0x47

Goroutine 62 (finished) created at:
  kubevirt.io/kubevirt/pkg/virt-api/rest.(*Streamer).Handle()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer.go:99 +0x776
  kubevirt.io/kubevirt/pkg/virt-api/rest.glob..func3.10.1()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:138 +0x22b
  net/http.HandlerFunc.ServeHTTP()
      /root/sdk/go1.18.5/src/net/http/server.go:2084 +0x4d
  net/http.serverHandler.ServeHTTP()
      /root/sdk/go1.18.5/src/net/http/server.go:2916 +0x896
  net/http.(*conn).serve()
      /root/sdk/go1.18.5/src/net/http/server.go:1966 +0xbaa
  net/http.(*Server).Serve.func3()
      /root/sdk/go1.18.5/src/net/http/server.go:3071 +0x58
==================
==================
WARNING: DATA RACE
Write at 0x00c000396a30 by goroutine 47:
  kubevirt.io/kubevirt/pkg/virt-api/rest.glob..func3.1()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:51 +0x23d
  github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func2()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:605 +0xe7

Previous read at 0x00c000396a30 by goroutine 55:
  kubevirt.io/kubevirt/pkg/virt-api/rest.glob..func3.1.5()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:72 +0x4e
  kubevirt.io/kubevirt/pkg/virt-api/rest.(*Streamer).Handle.func5()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer.go:100 +0x74

Goroutine 47 (running) created at:
  github.com/onsi/ginkgo/v2/internal.(*Suite).runNode()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:593 +0x8f9
  github.com/onsi/ginkgo/v2/internal.(*group).attemptSpec()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/group.go:181 +0x2c98
  github.com/onsi/ginkgo/v2/internal.(*group).run()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/group.go:303 +0xb9b
  github.com/onsi/ginkgo/v2/internal.(*Suite).runSpecs()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:304 +0xe71
  github.com/onsi/ginkgo/v2/internal.(*Suite).Run()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:83 +0x424
  github.com/onsi/ginkgo/v2.RunSpecs()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/core_dsl.go:280 +0x1044
  kubevirt.io/client-go/testutils.KubeVirtTestSuiteSetup()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/kubevirt.io/client-go/testutils/setup.go:55 +0x6a5
  kubevirt.io/kubevirt/pkg/virt-api/rest.TestRest()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/rest_suite_test.go:29 +0x2e
  testing.tRunner()
      /root/sdk/go1.18.5/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /root/sdk/go1.18.5/src/testing/testing.go:1486 +0x47

Goroutine 55 (finished) created at:
  kubevirt.io/kubevirt/pkg/virt-api/rest.(*Streamer).Handle()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer.go:100 +0x8f9
  kubevirt.io/kubevirt/pkg/virt-api/rest.glob..func3.10.1()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:138 +0x22b
  net/http.HandlerFunc.ServeHTTP()
      /root/sdk/go1.18.5/src/net/http/server.go:2084 +0x4d
  net/http.serverHandler.ServeHTTP()
      /root/sdk/go1.18.5/src/net/http/server.go:2916 +0x896
  net/http.(*conn).serve()
      /root/sdk/go1.18.5/src/net/http/server.go:1966 +0xbaa
  net/http.(*Server).Serve.func3()
      /root/sdk/go1.18.5/src/net/http/server.go:3071 +0x58
==================
•••
------------------------------
• [SLOW TEST] [5.002 seconds]
Streamer does not call keepAliveClient if the client connection upgrade failed
/root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:186
------------------------------
••••••••••==================
WARNING: DATA RACE
Read at 0x00c000872d90 by goroutine 38:
  runtime.chansend()
      /root/sdk/go1.18.5/src/runtime/chan.go:159 +0x0
  kubevirt.io/kubevirt/pkg/virt-api/rest.streamFuncResultChannelIsClosed.func1()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:393 +0x113

Previous write at 0x00c000872d90 by goroutine 30:
  runtime.closechan()
      /root/sdk/go1.18.5/src/runtime/chan.go:356 +0x0
  kubevirt.io/kubevirt/pkg/virt-api/rest.(*Streamer).Handle.func3()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer.go:97 +0x39
  runtime.deferreturn()
      /root/sdk/go1.18.5/src/runtime/panic.go:436 +0x32
  kubevirt.io/kubevirt/pkg/virt-api/rest.glob..func3.25.2()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:369 +0x2a6
  net/http.HandlerFunc.ServeHTTP()
      /root/sdk/go1.18.5/src/net/http/server.go:2084 +0x4d
  net/http.serverHandler.ServeHTTP()
      /root/sdk/go1.18.5/src/net/http/server.go:2916 +0x896
  net/http.(*conn).serve()
      /root/sdk/go1.18.5/src/net/http/server.go:1966 +0xbaa
  net/http.(*Server).Serve.func3()
      /root/sdk/go1.18.5/src/net/http/server.go:3071 +0x58

Goroutine 38 (running) created at:
  kubevirt.io/kubevirt/pkg/virt-api/rest.streamFuncResultChannelIsClosed()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:385 +0x15b
  kubevirt.io/kubevirt/pkg/virt-api/rest.glob..func3.25()
      /root/goapp/src/github.com/kubevirt/kubevirt/pkg/virt-api/rest/streamer_test.go:377 +0x610
  github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func2()
      /root/goapp/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:605 +0xe7

Goroutine 30 (finished) created at:
  net/http.(*Server).Serve()
      /root/sdk/go1.18.5/src/net/http/server.go:3071 +0x80c
  net/http/httptest.(*Server).goServe.func1()
      /root/sdk/go1.18.5/src/net/http/httptest/server.go:308 +0xb2
==================
send on closed channel
•••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••[{ "op": "test", "path": "/status/memoryDumpRequest", "value": null}, { "op": "add", "path": "/status/memoryDumpRequest", "value": {"claimName":"vol1","phase":"Associating"}}]
•[{ "op": "test", "path": "/status/memoryDumpRequest", "value": {"claimName":"vol1","phase":"Completed"}}, { "op": "replace", "path": "/status/memoryDumpRequest", "value": {"claimName":"vol1","phase":"Associating"}}]
•[{ "op": "test", "path": "/status/memoryDumpRequest", "value": {"claimName":"vol1","phase":"Failed"}}, { "op": "replace", "path": "/status/memoryDumpRequest", "value": {"claimName":"vol1","phase":"Associating"}}]
•
•
•
•
•
•[{ "op": "test", "path": "/status/memoryDumpRequest", "value": {"claimName":"vol1","phase":"Completed"}}, { "op": "replace", "path": "/status/memoryDumpRequest", "value": {"claimName":"vol1","phase":"Dissociating"}}]
•••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••

Ran 217 of 217 Specs in 6.529 seconds
SUCCESS! -- 217 Passed | 0 Failed | 0 Pending | 0 Skipped
    testing.go:1312: race detected during execution of test
--- FAIL: TestRest (6.58s)
=== CONT
    testing.go:1312: race detected during execution of test
FAIL
FAIL	kubevirt.io/kubevirt/pkg/virt-api/rest	6.752s
```
**What you expected to happen**:
success to run make go-test.

**How to reproduce it (as minimally and precisely as possible)**:
Steps to reproduce the behavior.

**Additional context**:
Add any other context about the problem here.

**Environment**:
- KubeVirt version (use `virtctl version`):  v0.54.0
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): centos8
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

--- a/pkg/virt-api/rest/BUILD.bazel
+++ b/pkg/virt-api/rest/BUILD.bazel
@@ -64,6 +64,8 @@ go_test(
         "expand_test.go",
         "profiler_test.go",
         "rest_suite_test.go",
+        "streamer_norace_test.go",
+        "streamer_race_test.go",
         "streamer_test.go",
         "subresource_test.go",
     ],

--- a/pkg/virt-api/rest/streamer_norace_test.go
+++ b/pkg/virt-api/rest/streamer_norace_test.go
@@ -1,0 +1,5 @@
+//go:build !race
+
+package rest
+
+const raceDetectorEnabled = false

--- a/pkg/virt-api/rest/streamer_race_test.go
+++ b/pkg/virt-api/rest/streamer_race_test.go
@@ -1,0 +1,5 @@
+//go:build race
+
+package rest
+
+const raceDetectorEnabled = true

--- a/pkg/virt-api/rest/streamer_test.go
+++ b/pkg/virt-api/rest/streamer_test.go
@@ -91,6 +91,10 @@ var _ = Describe("Streamer", func() {
 		validateVMICalled = false
 		dialCalled = false
 	})
+	AfterEach(func() {
+		emptyAndCloseChannel(streamToClientCalled)
+		emptyAndCloseChannel(streamToServerCalled)
+	})
 	It("fetches a VirtualMachineInstance", func() {
 		streamer.Handle(req, resp)
 		Expect(fetchVMICalled).To(BeTrue())
@@ -370,6 +374,12 @@ var _ = Describe("Streamer", func() {
 		Eventually(streamToServerCalled, defaultTestTimeout).Should(Receive())
 	})
 	It("closes the result channel after both streams have returned", func() {
+		// This test is mutually exclusive with the `-race` flag, as there is no other way to check
+		// if a write-only channel is closed than to write to it and catch the panic.
+		if raceDetectorEnabled {
+			Skip("Data Race Detector is enabled")
+		}
+
 		var results chan<- streamFuncResult
 		streamer.streamToClient = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
 			result <- goerrors.New("done")
@@ -410,6 +420,14 @@ func streamFuncResultChannelIsClosed(channel chan<- streamFuncResult, timeout ti
 	}()
 
 	return <-closed
+}
+
+func emptyAndCloseChannel(channel chan struct{}) {
+	select {
+	case <-channel:
+	default:
+	}
+	close(channel)
 }
 
 func testWebsocketDial(handler http.HandlerFunc) (*httptest.Server, *websocket.Conn, *http.Response, error) {

--- a/pkg/virt-api/rest/streamer_test.go
+++ b/pkg/virt-api/rest/streamer_test.go
@@ -9,6 +9,7 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"strings"
+	"sync"
 	"time"
 
 	restful "github.com/emicklei/go-restful/v3"
@@ -148,13 +149,17 @@ var _ = Describe("Streamer", func() {
 		Expect(dialCalled).To(BeFalse())
 	})
 	It("upgrades the client connection", func() {
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, wsResp, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
 			streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 		}))
 		Expect(err).NotTo(HaveOccurred())
 		defer srv.Close()
 		Expect(wsResp.StatusCode).To(Equal(101))
 		defer ws.Close()
+		wg.Wait()
 	})
 	It("does not attempt the client connection upgrade on a failed dial", func() {
 		directDialer.dial = mockDialer{
@@ -162,17 +167,26 @@ var _ = Describe("Streamer", func() {
 				return nil, errors.NewInternalError(goerrors.New("test error"))
 			},
 		}
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, _, wsResp, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).To(HaveOccurred())
 		}))
 		Expect(err).To(HaveOccurred())
 		defer srv.Close()
 		Expect(wsResp.StatusCode).To(Equal(http.StatusInternalServerError))
+		wg.Wait()
 	})
 	Context("clientConnectionUpgrade", func() {
 		It("does not fail the upgrade on a correct request", func() {
+			var wg sync.WaitGroup
+			wg.Add(1)
 			srv, ws, wsResp, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+				defer wg.Done()
+				defer GinkgoRecover()
 				_, upgradeErr := clientConnectionUpgrade(restful.NewRequest(r), restful.NewResponse(rw))
 				Expect(upgradeErr).NotTo(HaveOccurred())
 			}))
@@ -180,6 +194,7 @@ var _ = Describe("Streamer", func() {
 			defer srv.Close()
 			Expect(wsResp.StatusCode).To(Equal(101))
 			defer ws.Close()
+			wg.Wait()
 		})
 	})
 	It("calls keepAliveClient if set", func() {
@@ -190,7 +205,11 @@ var _ = Describe("Streamer", func() {
 			Expect(ctx).NotTo(BeNil())
 			Expect(conn).NotTo(BeNil())
 		}
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).NotTo(HaveOccurred())
 		}))
@@ -198,6 +217,7 @@ var _ = Describe("Streamer", func() {
 		defer srv.Close()
 		defer ws.Close()
 		Eventually(call, defaultTestTimeout).Should(Receive())
+		wg.Wait()
 	})
 	It("does not call keepAliveClient if the client connection upgrade failed", func() {
 		call := make(chan struct{})
@@ -209,12 +229,17 @@ var _ = Describe("Streamer", func() {
 	})
 	It("does start streamToClient with connections", func() {
 		streamer.streamToClient = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			defer GinkgoRecover()
 			Expect(clientSocket).NotTo(BeNil())
 			Expect(serverConn).NotTo(BeNil())
 			result <- nil
 			streamToClientCalled <- struct{}{}
 		}
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).NotTo(HaveOccurred())
 		}))
@@ -222,15 +247,21 @@ var _ = Describe("Streamer", func() {
 		defer srv.Close()
 		defer ws.Close()
 		Eventually(streamToClientCalled, defaultTestTimeout).Should(Receive())
+		wg.Wait()
 	})
 	It("does start streamToServer with connections", func() {
 		streamer.streamToServer = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+			defer GinkgoRecover()
 			Expect(clientSocket).NotTo(BeNil())
 			Expect(serverConn).NotTo(BeNil())
 			result <- nil
 			streamToServerCalled <- struct{}{}
 		}
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).NotTo(HaveOccurred())
 		}))
@@ -238,9 +269,14 @@ var _ = Describe("Streamer", func() {
 		defer srv.Close()
 		defer ws.Close()
 		Eventually(streamToServerCalled, defaultTestTimeout).Should(Receive())
+		wg.Wait()
 	})
 	It("closes clientSocket when streamToClient returns", func() {
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).NotTo(HaveOccurred())
 		}))
@@ -250,9 +286,14 @@ var _ = Describe("Streamer", func() {
 		ws.SetReadDeadline(time.Now().Add(defaultTestTimeout))
 		_, _, err = ws.ReadMessage()
 		Expect(err).To(BeAssignableToTypeOf(&websocket.CloseError{}))
+		wg.Wait()
 	})
 	It("closes serverSocket when streamToServer returns", func() {
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).NotTo(HaveOccurred())
 		}))
@@ -262,18 +303,27 @@ var _ = Describe("Streamer", func() {
 		serverPipe.SetReadDeadline(time.Now().Add(defaultTestTimeout))
 		_, err = serverPipe.Read([]byte{})
 		Expect(err).To(Equal(io.EOF))
+		wg.Wait()
 	})
 	It("closes clientSocket when keepAliveClient cancels context", func() {
 		streamer.streamToClient = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
 			streamToClientCalled <- struct{}{}
+			time.Sleep(defaultTestTimeout * 2)
+			result <- nil
 		}
 		streamer.streamToServer = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
 			streamToServerCalled <- struct{}{}
+			time.Sleep(defaultTestTimeout * 2)
+			result <- nil
 		}
 		streamer.keepAliveClient = func(ctx context.Context, conn *websocket.Conn, cancel func()) {
 			cancel()
 		}
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).NotTo(HaveOccurred())
 		}))
@@ -283,13 +333,14 @@ var _ = Describe("Streamer", func() {
 		ws.SetReadDeadline(time.Now().Add(defaultTestTimeout))
 		_, _, err = ws.ReadMessage()
 		Expect(err).To(BeAssignableToTypeOf(&websocket.CloseError{}))
+		wg.Wait()
 	})
-	It("cleans up the streamToClient goroutine on termination", func() {
-		streamer.streamToClient = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
-			result <- nil
-			streamToClientCalled <- struct{}{}
-		}
+	It("calls the streamToClient goroutine", func() {
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).NotTo(HaveOccurred())
 		}))
@@ -297,13 +348,14 @@ var _ = Describe("Streamer", func() {
 		defer srv.Close()
 		defer ws.Close()
 		Eventually(streamToClientCalled, defaultTestTimeout).Should(Receive())
+		wg.Wait()
 	})
-	It("cleans up the streamToServer goroutine on termination", func() {
-		streamer.streamToServer = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
-			result <- nil
-			streamToServerCalled <- struct{}{}
-		}
+	It("calls the streamToServer goroutine", func() {
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).NotTo(HaveOccurred())
 		}))
@@ -311,12 +363,9 @@ var _ = Describe("Streamer", func() {
 		defer srv.Close()
 		defer ws.Close()
 		Eventually(streamToServerCalled, defaultTestTimeout).Should(Receive())
+		wg.Wait()
 	})
 	It("starts to cleanup after the first stream returns", func() {
-		streamer.streamToClient = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
-			result <- nil
-			streamToClientCalled <- struct{}{}
-		}
 		streamer.streamToServer = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
 			defer GinkgoRecover()
 			serverConn.SetReadDeadline(time.Now().Add(defaultTestTimeout))
@@ -326,6 +375,7 @@ var _ = Describe("Streamer", func() {
 			streamToServerCalled <- struct{}{}
 		}
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).NotTo(HaveOccurred())
 		}))
@@ -343,8 +393,13 @@ var _ = Describe("Streamer", func() {
 		}
 		streamer.streamToServer = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
 			streamToServerCalled <- struct{}{}
+			time.Sleep(defaultTestTimeout * 2)
+			result <- nil
 		}
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
 			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).To(Equal(testErrStreamEnded))
@@ -353,17 +408,23 @@ var _ = Describe("Streamer", func() {
 		defer srv.Close()
 		defer ws.Close()
 		Eventually(streamToClientCalled, defaultTestTimeout).Should(Receive())
+		wg.Wait()
 	})
 	It("returns the first stream result/error if streamToServer terminates", func() {
 		testErrStreamEnded := goerrors.New("stream ended")
 		streamer.streamToClient = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
 			streamToClientCalled <- struct{}{}
+			time.Sleep(defaultTestTimeout * 2)
+			result <- nil
 		}
 		streamer.streamToServer = func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
 			result <- testErrStreamEnded
 			streamToServerCalled <- struct{}{}
 		}
+		var wg sync.WaitGroup
+		wg.Add(1)
 		srv, ws, _, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			defer wg.Done()
 			defer GinkgoRecover()
 			handleErr := streamer.Handle(restful.NewRequest(r), restful.NewResponse(rw))
 			Expect(handleErr).To(Equal(testErrStreamEnded))
@@ -372,6 +433,7 @@ var _ = Describe("Streamer", func() {
 		defer srv.Close()
 		defer ws.Close()
 		Eventually(streamToServerCalled, defaultTestTimeout).Should(Receive())
+		wg.Wait()
 	})
 	It("closes the result channel after both streams have returned", func() {
 		// This test is mutually exclusive with the `-race` flag, as there is no other way to check

